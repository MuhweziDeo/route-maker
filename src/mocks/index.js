export const locations = [
    {
        geometry: {
            lat: 0.400,
            lng: 0.344
        },
        formatted: "Kampala"
    },
    {
        geometry: {
            lat: 0.400,
            lng: 0.344
        },
        formatted: "Ntinda"
    },
    {
        geometry: {
            lat: 0.400,
            lng: 0.344
        },
        formatted: "Kiwatule"
    }
];
