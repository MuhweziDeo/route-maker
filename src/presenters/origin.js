import React from "react";
import {Button, Text, View, TouchableOpacity, StyleSheet} from "react-native";
import Autocomplete from "react-native-autocomplete-input";


//TODO enable user select current location as origin

export const OriginPresenter = ({navigation, origin,originResults, setOrigin,
                                    setOriginResults, filterOrigin }) => {
    return (
        <View style={{ width: 400,height: 900,
            paddingTop: 10,
            backgroundColor: '#F5FCFF',
            paddingLeft: 20, flex: 1 }}>
            <View style={{ width:300, marginBottom: 10}}>
                <Button disabled={origin === {} || !origin.geometry} title={"Next"}
                        onPress={() => navigation.navigate('Destination')}/>
            </View>
            <Autocomplete
                data={originResults}
                defaultValue={origin.formatted}
                onChangeText={text => filterOrigin(text)}
                placeholder={"Enter Origin"}
                containerStyle={styles.autoCompleteContainer}
                renderItem={({ item, i }) => (
                    <TouchableOpacity key={i} style={styles.itemText} onPress={() => {
                     setOrigin(item);
                     setOriginResults([])
                    }}>
                        <Text>{item.formatted}</Text>
                    </TouchableOpacity>
                )}
            />
        </View>
    );

};


export const styles = StyleSheet.create({
    itemText: {
        fontSize: 15,
        margin: 2,
        padding: 5,
        backgroundColor: "#F5FCFF",
        color: "white"
    },
    autoCompleteContainer: {
        width: 320
    }
});
