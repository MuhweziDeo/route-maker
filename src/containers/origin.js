import React, { useState, useEffect, useContext} from "react";
import { OriginPresenter } from "../presenters";
import geo from "opencage-api-client";
import {openCageApiKey} from "../config";
import {AppContext} from "../../App";

export const OriginContainer = props => {
    const context = useContext(AppContext);
    const[origin, setOrigin] = useState(context.origin || {});
    const[originResults, setOriginResults] = useState([]);

    const filterOrigin = async (value) => {
      setOrigin(value);
      const {results} = await geo.geocode({q: value.toLowerCase(), key: openCageApiKey});
      setOriginResults(results);
    };

    useEffect(()=> {
        if(origin === {}) {
            setOriginResults([]);
        }
        context.origin = origin;
    }, [origin]);

    return (
        <OriginPresenter
            {...props}
            origin={origin}
            setOrigin={setOrigin}
            originResults={originResults}
            setOriginResults={setOriginResults}
            filterOrigin={filterOrigin}
        />
    )
};
