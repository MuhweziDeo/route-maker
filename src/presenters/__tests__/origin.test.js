import React from "react";
import renderer from 'react-test-renderer'
import { OriginPresenter } from "../origin";
import {Button, TouchableOpacity} from "react-native";;
import {locations} from "../../mocks";
import Autocomplete from "react-native-autocomplete-input";

export const OriginComponentProps = {
    navigation: {
        navigate: jest.fn()
    },
    origin: {...locations[0]},
    originResults: [...locations],
    setOrigin: jest.fn(),
    setOriginResults: jest.fn(),
    filterOrigin: jest.fn()
};

let wrapper;
let instance;

describe('<OriginPresenter/>', () => {
    beforeEach(() => {
        wrapper = renderer.create(<OriginPresenter {...OriginComponentProps}/>);
        instance = wrapper.root;
    });
    it("should match snapshot", () => {
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it("should disable button if origin is not provided", () => {

        const button = instance.findByType(Button);
        expect(button.props.disabled).toBeFalsy()
    });

    it("should navigate on click next", () => {

        const button = instance.findByType(Button);
        button.props.onPress();
        expect(OriginComponentProps.navigation.navigate).toHaveBeenCalled();
    });

    it("set origin on click item", () => {
        let TouchableOpacityElement = instance.findAllByType(TouchableOpacity);
        TouchableOpacityElement[1].props.onPress();
        expect(OriginComponentProps.setOrigin).toHaveBeenCalled();
        expect(OriginComponentProps.setOriginResults).toHaveBeenCalled();
    });

    it("should filter origin on AutoComplete change", () => {
        let AutoCompleteElement = instance.findByType(Autocomplete);
        AutoCompleteElement.props.onChangeText("kampala");
        expect(OriginComponentProps.filterOrigin).toHaveBeenCalled();
    });
});
