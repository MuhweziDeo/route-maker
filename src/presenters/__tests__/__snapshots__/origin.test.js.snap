// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OriginPresenter/> should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "height": 900,
      "paddingLeft": 20,
      "paddingTop": 10,
      "width": 400,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 10,
        "width": 300,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          disabled={false}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Next
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "zIndex": 1,
        },
        Object {
          "width": 320,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "#b9b9b9",
            "borderRadius": 1,
            "borderWidth": 1,
          },
          undefined,
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        containerStyle={
          Object {
            "width": 320,
          }
        }
        data={
          Array [
            Object {
              "formatted": "Kampala",
              "geometry": Object {
                "lat": 0.4,
                "lng": 0.344,
              },
            },
            Object {
              "formatted": "Ntinda",
              "geometry": Object {
                "lat": 0.4,
                "lng": 0.344,
              },
            },
            Object {
              "formatted": "Kiwatule",
              "geometry": Object {
                "lat": 0.4,
                "lng": 0.344,
              },
            },
          ]
        }
        defaultValue="Kampala"
        flatListProps={Object {}}
        keyboardShouldPersistTaps="always"
        onChangeText={[Function]}
        onEndEditing={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        placeholder="Enter Origin"
        rejectResponderTermination={true}
        renderItem={[Function]}
        renderSeparator={null}
        renderTextInput={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "height": 40,
              "paddingLeft": 3,
            },
            undefined,
          ]
        }
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      onStartShouldSetResponderCapture={[Function]}
    >
      <RCTScrollView
        data={
          Array [
            Object {
              "formatted": "Kampala",
              "geometry": Object {
                "lat": 0.4,
                "lng": 0.344,
              },
            },
            Object {
              "formatted": "Ntinda",
              "geometry": Object {
                "lat": 0.4,
                "lng": 0.344,
              },
            },
            Object {
              "formatted": "Kiwatule",
              "geometry": Object {
                "lat": 0.4,
                "lng": 0.344,
              },
            },
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="always"
        maxToRenderPerBatch={10}
        numColumns={1}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        renderSeparator={null}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderColor": "#b9b9b9",
              "borderRadius": 1,
              "borderTopWidth": 0,
              "borderWidth": 1,
              "left": 0,
              "position": "absolute",
              "right": 0,
            },
            undefined,
          ]
        }
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#F5FCFF",
                  "color": "white",
                  "fontSize": 15,
                  "margin": 2,
                  "opacity": 1,
                  "padding": 5,
                }
              }
            >
              <Text>
                Kampala
              </Text>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#F5FCFF",
                  "color": "white",
                  "fontSize": 15,
                  "margin": 2,
                  "opacity": 1,
                  "padding": 5,
                }
              }
            >
              <Text>
                Ntinda
              </Text>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#F5FCFF",
                  "color": "white",
                  "fontSize": 15,
                  "margin": 2,
                  "opacity": 1,
                  "padding": 5,
                }
              }
            >
              <Text>
                Kiwatule
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
